In [1]:
%pip install transformers torch tf-keras

Note: you may need to restart the kernel to use updated packages.


Bài 1: Khôi phục Masked Token (Masked Language Modeling) Trong tác vụ
này, chúng ta sẽ che một vài từ trong câu bằng một token đặc biệt (thường là [MASK])
và yêu cầu mô hình dự đoán xem từ gốc là gì. Chúng ta sẽ sử dụng một mô hình thuộc
họ BERT.

Yêu cầu: Sử dụng pipeline fill-mask để dự đoán từ bị thiếu trong câu sau: Hanoi is
the [MASK] of Vietnam.

In [4]:
from transformers import pipeline
# 1. Tải pipeline "fill-mask"
# Pipeline này sẽ tự động tải một mô hình mặc định phù hợp (thường là một biến thể của BERT)
mask_filler = pipeline("fill-mask")
# 2. Câu đầu vào với token [MASK]
input_sentence = "Hanoi is the <mask> of Vietnam."
# 3. Thực hiện dự đoán
# top_k=5 yêu cầu mô hình trả về 5 dự đoán hàng đầu
predictions = mask_filler(input_sentence, top_k=5)
# 4. In kết quả
print(f"Câu gốc: {input_sentence}")
for pred in predictions:
    print(f"Dự đoán: '{pred['token_str']}' với độ tin cậy: {pred['score']:.4f}")
    print(f" -> Câu hoàn chỉnh: {pred['sequence']}")

No model was supplied, defaulted to distilbert/distilroberta-base and revision fb53ab8 (https://huggingface.co/distilbert/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.
Some weights of the model checkpoint at distilbert/distilroberta-base were not used when initializing RobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cpu


Câu gốc: Hanoi is the <mask> of Vietnam.
Dự đoán: ' capital' với độ tin cậy: 0.9341
 -> Câu hoàn chỉnh: Hanoi is the capital of Vietnam.
Dự đoán: ' Republic' với độ tin cậy: 0.0300
 -> Câu hoàn chỉnh: Hanoi is the Republic of Vietnam.
Dự đoán: ' Capital' với độ tin cậy: 0.0105
 -> Câu hoàn chỉnh: Hanoi is the Capital of Vietnam.
Dự đoán: ' birthplace' với độ tin cậy: 0.0054
 -> Câu hoàn chỉnh: Hanoi is the birthplace of Vietnam.
Dự đoán: ' heart' với độ tin cậy: 0.0014
 -> Câu hoàn chỉnh: Hanoi is the heart of Vietnam.


Khó khăn:
- Mô hình được pipeline của transformer lựa chọn không sử dụng [MASK] mà sử dụng <mask>

Câu hỏi:
1. Mô hình đã dự đoán đúng từ capital không?
    - Mô hình đã dự đoán đúng từ "capital" với độ tin cậy lên đến 0.9341
2. Tại sao các mô hình Encoder-only như BERT lại phù hợp cho tác vụ này?
    - Mô hình BERT phù hợp cho tác vụ đoán từ bị mask vì BERT sử dụng mô hình MLM = Masked Language Modelling, tức là tác vụ đoán từ bị mask chính là cách BERT được train.
    - Cụ thể, các Encoder có cơ chế self-attention 2 chiều dể tập trung cả hai hướng trước sau, không bị ảnh hưởng bởi thứ tự sinh token.

Bài 2: Dự đoán từ tiếp theo (Next Token Prediction) Tác vụ này yêu cầu mô
hình sinh ra phần tiếp theo của một đoạn văn bản cho trước. Chúng ta sẽ sử dụng một
mô hình thuộc họ GPT.

Yêu cầu: Sử dụng pipeline text-generation để sinh ra phần tiếp theo cho câu: The best
thing about learning NLP is

In [5]:
from transformers import pipeline
# 1. Tải pipeline "text-generation"
# Pipeline này sẽ tự động tải một mô hình phù hợp (thường là GPT-2)
generator = pipeline("text-generation")
# 2. Đoạn văn bản mồi
prompt = "The best thing about learning NLP is"
# 3. Sinh văn bản
# max_length: tổng độ dài của câu mồi và phần được sinh ra
# num_return_sequences: số lượng chuỗi kết quả muốn nhận
generated_texts = generator(prompt, max_length=50, num_return_sequences=1)
# 4. In kết quả
print(f"Câu mồi: '{prompt}'")
for text in generated_texts:
    print("Văn bản được sinh ra:")
    print(text['generated_text'])

No model was supplied, defaulted to openai-community/gpt2 and revision 607a30d (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.
d:\College\year6s1\nlp-dl\.venv\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\ADMIN\.cache\huggingface\hub\models--openai-community--gpt2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://do

Câu mồi: 'The best thing about learning NLP is'
Văn bản được sinh ra:
The best thing about learning NLP is that it's a matter of practice. In theory, it could be done in three steps, but there are ways of doing it that are much more interesting and take much more time.

Learn to Recognize Empathy

In a nutshell, you want to learn to recognize empathy or empathy for others.

In a way, empathy is just one way to see the world. It's all about feeling good about yourself.

When you're feeling good about yourself, you can always feel bad about yourself. This is why I love to be able to tell people that I'm a good person because I'm feeling good about myself.

Now, you can also see how you can see both the world and yourself.

I think that we all need to learn to see each other as individuals.

I think that people need to be able to see the world as individuals, and they need to feel good about themselves.

I think that people need to be able to see the world as individuals, and they need to

Câu hỏi:
1. Kết quả sinh ra có hợp lý không?
    - Kết quả đoạn đầu tiên hợp lý nhưng sau đó nói về Empathy có vẻ không còn hợp lý.
2. Tại sao các mô hình Decoder-only như GPT lại phù hợp cho tác vụ này?
    - Decoder-only được huấn luyện theo cách dự đoán từ tiếp theo bằng các từ trước đó
    - Decoder sinh tuần tự
    - Decoder chỉ nhìn bên trái không phải hai chiều

Bài 3: Tính toán Vector biểu diễn của câu (Sentence Representation) Một
trong những ứng dụng mạnh mẽ nhất của BERT là khả năng chuyển đổi một câu thành
một vector số có chiều dài cố định, nắm bắt được ngữ nghĩa của câu đó. Vector này
có thể được dùng cho các tác vụ khác như phân loại, tìm kiếm tương đồng, v.v.

Có hai cách phổ biến để lấy vector biểu diễn cho cả câu:
1. Lấy vector đầu ra của token [CLS] (token đặc biệt được thêm vào đầu mỗi câu).
2. Lấy trung bình cộng của các vector đầu ra của tất cả các token trong câu (Mean
Pooling). Cách này thường cho kết quả tốt hơn.

Yêu cầu: Viết code để tính toán vector biểu diễn cho câu This is a sample sentence.
bằng phương pháp Mean Pooling.

In [7]:
import torch
from transformers import AutoTokenizer, AutoModel

# 1. Chọn một mô hình BERT
model_name = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name)
# 2. Câu đầu vào
sentences = ["This is a sample sentence."]
# 3. Tokenize câu
# padding=True: đệm các câu ngắn hơn để có cùng độ dài
# truncation=True: cắt các câu dài hơn
# return_tensors='pt': trả về kết quả dưới dạng PyTorch tensors
inputs = tokenizer(sentences, padding=True, truncation=True, return_tensors='pt')
# 4. Đưa qua mô hình để lấy hidden states
# torch.no_grad() để không tính toán gradient, tiết kiệm bộ nhớ
with torch.no_grad():
    outputs = model(**inputs)
# outputs.last_hidden_state chứa vector đầu ra của tất cả các token
last_hidden_state = outputs.last_hidden_state
# shape: (batch_size, sequence_length, hidden_size)
# 5. Thực hiện Mean Pooling
# Để tính trung bình chính xác, chúng ta cần bỏ qua các token đệm (padding tokens)
attention_mask = inputs['attention_mask']
mask_expanded = attention_mask.unsqueeze(-1).expand(last_hidden_state.size()).float()
sum_embeddings = torch.sum(last_hidden_state * mask_expanded, 1)
sum_mask = torch.clamp(mask_expanded.sum(1), min=1e-9)
sentence_embedding = sum_embeddings / sum_mask
# 6. In kết quả
print("Vector biểu diễn của câu:")
print(sentence_embedding)
print("\nKích thước của vector:", sentence_embedding.shape)

d:\College\year6s1\nlp-dl\.venv\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\ADMIN\.cache\huggingface\hub\models--bert-base-uncased. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)
Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. 

Vector biểu diễn của câu:
tensor([[-6.3875e-02, -4.2837e-01, -6.6779e-02, -3.8430e-01, -6.5785e-02,
         -2.1826e-01,  4.7636e-01,  4.8659e-01,  4.0036e-05, -7.4274e-02,
         -7.4740e-02, -4.7635e-01, -1.9773e-01,  2.4824e-01, -1.2162e-01,
          1.6679e-01,  2.1045e-01, -1.4576e-01,  1.2637e-01,  1.8635e-02,
          2.4640e-01,  5.7090e-01, -4.7014e-01,  1.3782e-01,  7.3650e-01,
         -3.3808e-01, -5.0329e-02, -1.6453e-01, -4.3517e-01, -1.2900e-01,
          1.6516e-01,  3.4004e-01, -1.4930e-01,  2.2422e-02, -1.0488e-01,
         -5.1916e-01,  3.2964e-01, -2.2162e-01, -3.4206e-01,  1.1993e-01,
         -7.0148e-01, -2.3126e-01,  1.1224e-01,  1.2550e-01, -2.5191e-01,
         -4.6375e-01, -2.7261e-02, -2.8415e-01, -9.9249e-02, -3.7019e-02,
         -8.9192e-01,  2.5005e-01,  1.5816e-01,  2.2701e-01, -2.8497e-01,
          4.5300e-01,  5.0921e-03, -7.9441e-01, -3.1008e-01, -1.7403e-01,
          4.3029e-01,  1.6816e-01,  1.0590e-01, -4.8987e-01,  3.1856e-01,
          3.

Câu hỏi:
1. Kích thước (chiều) của vector biểu diễn là bao nhiêu? Con số này tương ứng với tham số nào của mô hình BERT?
    - Kích thước của vector biểu diễn là [1,768]
    - Con số này tương ứng với tham số Hidden embedding size chính là số chiều vector mà bert sử dụng để biểu diễn embedding khi token đi qua các encoder layer.

2. Tại sao chúng ta cần sử dụng attention_mask khi thực hiện Mean Pooling?
    - Mean-Pooling là lấy mean embedding của các token
    - Các token là các câu có thể có độ dài ngắn khác nhau => Cần có padding
    - attention-masks là phương pháp bỏ qua các padding khi áp dụng mean pooling
    - Vậy ta cần sử dụng attention-mask khi thực hiện mean pooling vì các padding với các giá trị 0 có thể làm lệch giá trị mean tính được.